tiledb/api/src/group/mod.rs

```diff
@@ -0,0 +1,745 @@
+        &self,
+        config: &Config,
+        group_uri: S,
+    ) -> TileDBResult<()>
+    where
+        S: AsRef<str>,
+    {
+        let c_context = self.context.capi();
+        let c_group_uri = cstring!(group_uri.as_ref());
+        let cfg = config.capi();
+        self.context.capi_return(unsafe {
+            ffi::tiledb_group_consolidate_metadata(
+                c_context,
+                c_group_uri.as_ptr(),
+                cfg,
+            )
+        })?;
+        Ok(())
+    }
+
+    pub fn vacuum_metadata<S>(
+        &self,
+        config: &Config,
+        group_uri: S,
+    ) -> TileDBResult<()>
+    where
+        S: AsRef<str>,
+    {
+        let c_context = self.context.capi();
+        let c_group_uri = cstring!(group_uri.as_ref());
+        let cfg = config.capi();
+        self.context.capi_return(unsafe {
+            ffi::tiledb_group_vacuum_metadata(
+                c_context,
+                c_group_uri.as_ptr(),
+                cfg,
+            )
+        })?;
+        Ok(())
+    }
+}
+
+impl Drop for Group<'_> {
+    fn drop(&mut self) {
+        let c_context = self.context.capi();
+        let c_group = Self::capi(self);
+
+        let mut c_open: i32 = out_ptr!();
+        self.context
+            .capi_return(unsafe {
+                ffi::tiledb_group_is_open(c_context, c_group, &mut c_open)
+            })
+            .expect("TileDB internal error when checking for open group.");
+        if c_open > 0 {
+            self.context
+                .capi_return(unsafe {
+                    ffi::tiledb_group_close(c_context, c_group)
+                })
+                .expect("TileDB internal error when closing group");
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        array::Array,
+        array::ArrayType,
+        context::Context,
+        datatype::Datatype,
+        error::Error,
+        group::{Group, QueryType},
+        key::LookupKey,
+        metadata::{self, Metadata},
+    };
+    use crate::{
+        array::{
+            AttributeBuilder, CellOrder, Dimension, DimensionBuilder,
+            DomainBuilder, SchemaBuilder, TileOrder,
+        },
+        context::ObjectType,
+        Result as TileDBResult,
+    };
+    use tempfile::TempDir;
+    #[test]
+    fn test_group_metadata() -> TileDBResult<()> {
+        let tmp_dir =
+            TempDir::new().map_err(|e| Error::Other(e.to_string()))?;
+
+        let tdb = Context::new()?;
+        let group1_path = tmp_dir.path().join("group1");
+        let group1_uri = group1_path.to_str().unwrap();
+        Group::create(&tdb, group1_uri)?;
+
+        let mut group1_err =
+            Group::open(&tdb, group1_uri.to_owned(), QueryType::Read)?;
+        let res = group1_err.put_metadata(Metadata::new(
+            "key".to_owned(),
+            Datatype::Int32,
+            vec![5],
+        ));
+        assert!(res.is_err());
+
+        std::mem::drop(group1_err);
```

Comment:
I suppose it's something of a stylistic preference, but you can avoid this function call by using inner blocks instead.
```
{
    let group1_err = Group::open(&tdb, group1_uri.to_owned(), QueryType::Read)?;
    let res = group1_err.put_metadata(...);
    ...
    /* `res` will be dropped here, then `group1_err` */
}
```
I prefer this myself as it will drop everything used for this part of the test and be a better visual cue via indentation.